src/Dialect/ONNX/ONNXOps/RNN/RNN.cpp

```diff
@@ -22,138 +22,162 @@ using namespace onnx_mlir;
-    hiddenSize = op->hidden_size().value();
+  IndexExpr hiddenSize;
+  if (operandAdaptor.hidden_size().has_value()) {
+    hiddenSize = LiteralIndexExpr(operandAdaptor.hidden_size().value());
   } else {
     // Infer hidden_size from wShape and rShape if possible.
-    if (!ShapedType::isDynamic(rShape[2]))
-      hiddenSize = rShape[2];
-    else if (!ShapedType::isDynamic(rShape[1]))
-      hiddenSize = rShape[1] / gates;
-    else if (!ShapedType::isDynamic(wShape[1]))
-      hiddenSize = wShape[1] / gates;
+    if (rDims[2].isLiteral())
+      hiddenSize = rDims[2];
+    else if (rDims[1].isLiteral())
+      hiddenSize = rDims[1].floorDiv(gates);
+    else if (wDims[1].isLiteral())
+      hiddenSize = wDims[1].floorDiv(gates);
+    else
+      // Pick one of the option above.
+      hiddenSize = rDims[2];
     // Update hidden_size attribute.
-    if (!ShapedType::isDynamic(hiddenSize)) {
+    if (hiddenSize.isLiteral()) {
       auto builder = Builder(op->getContext());
       auto hiddenSizeAttr =
           IntegerAttr::get(builder.getIntegerType(64, /*isSigned=*/true),
-              APInt(64, /*value=*/hiddenSize, /*isSigned=*/true));
-      op->hidden_sizeAttr(hiddenSizeAttr);
+              APInt(64, /*value=*/hiddenSize.getLiteral(), /*isSigned=*/true));
+      rnnOp.hidden_sizeAttr(hiddenSizeAttr);
     }
   }
 
   // Get direction.
-  int64_t numDir;
-  if ((op->direction() == "forward") || (op->direction() == "reverse"))
-    numDir = 1;
-  else if (op->direction() == "bidirectional")
-    numDir = 2;
+  IndexExpr numDir;
+  if ((operandAdaptor.direction() == "forward") ||
+      (operandAdaptor.direction() == "reverse"))
+    numDir = LiteralIndexExpr(1);
+  else if (operandAdaptor.direction() == "bidirectional")
+    numDir = LiteralIndexExpr(2);
   else
     return op->emitError(
         "direction attribute must be one of the strings: forward, "
         "reverse, and bidirectional");
 
   // Set result types. There are always 2 (RNN, GRU) or 3 results
   // but they are sometimes optional in which case they have NoneType.
-  assert((op->getNumResults() == 2 || op->getNumResults() == 3) &&
+  assert((rnnOp->getNumResults() == 2 || rnnOp->getNumResults() == 3) &&
          "RNN, GRU have 2 results, LSTM has 3");
+
   // Y :: [batch_size, seq_length, num_dir, hidden_size] if batchwiseLayout
   // Y :: [seq_length, num_dir, batch_size, hidden_size] otherwise
+  DimsExpr yOutputDims;
   Type yTy = op->getResult(0).getType();
   if (!yTy.isa<NoneType>()) {
     if (batchwiseLayout) {
-      yTy = RankedTensorType::get(
-          {batchSize, seqLength, numDir, hiddenSize}, elementType);
+      yOutputDims = {batchSize, seqLength, numDir, hiddenSize};
     } else {
-      yTy = RankedTensorType::get(
-          {seqLength, numDir, batchSize, hiddenSize}, elementType);
+      yOutputDims = {seqLength, numDir, batchSize, hiddenSize};
     }
-    op->getResult(0).setType(yTy);
+    IndexExpr::debugPrint("set0", yOutputDims);
```

Comment:
Not sure if you still want to keep this debug in the final code. Same for `set1` and `set2` below.

Reply:
should have caught it, thanks for spotting it.